src/Ledger/Deleg.lagda

```diff
@@ -245,43 +254,51 @@ instance
     (ccreghot c _) _ (GOVCERT-ccreghot ¬p)
     rewrite dec-no ((c , nothing) ∈? (ccKeys ˢ)) ¬p = refl
 
-instance
   Computational-CERT : Computational _⊢_⇀⦇_,CERT⦈_
-  Computational-CERT .computeProof Γ@(⟦ e , pp , vs ⟧ᶜ) ⟦ stᵈ , stᵖ , stᵍ ⟧ᶜ dCert
-    with computeProof pp stᵈ dCert | computeProof pp stᵖ dCert | computeProof Γ stᵍ dCert
+  Computational-CERT .computeProof Γ@(⟦ e , pp , vs , _ ⟧ᶜ) ⟦ stᵈ , stᵖ , stᵍ ⟧ᶜˢ dCert
+    with computeProof ⟦ pp , PState.pools stᵖ ⟧ᵈᵉ stᵈ dCert
+       | computeProof pp stᵖ dCert | computeProof Γ stᵍ dCert
   ... | just (_ , h) | _            | _            = just (-, CERT-deleg h)
   ... | nothing      | just (_ , h) | _            = just (-, CERT-pool h)
   ... | nothing      | nothing      | just (_ , h) = just (-, CERT-vdel h)
   ... | nothing      | nothing      | nothing      = nothing
-  Computational-CERT .completeness ⟦ _ , pp , _ ⟧ᶜ ⟦ stᵈ , stᵖ , stᵍ ⟧ᶜ
-    dCert@(delegate c mv mc d) ⟦ stᵈ' , stᵖ , stᵍ ⟧ᶜ (CERT-deleg h)
-    with computeProof pp stᵈ dCert | completeness _ _ _ _ h
+  Computational-CERT .completeness ⟦ _ , pp , _ , wdrls ⟧ᶜ ⟦ stᵈ , stᵖ , stᵍ ⟧ᶜˢ
+    dCert@(delegate c mv mc d) ⟦ stᵈ' , stᵖ , stᵍ ⟧ᶜˢ (CERT-deleg h)
+    with computeProof ⟦ pp , PState.pools stᵖ ⟧ᵈᵉ stᵈ dCert | completeness _ _ _ _ h
   ... | just _ | refl = refl
-  Computational-CERT .completeness ⟦ _ , pp , _ ⟧ᶜ ⟦ stᵈ , stᵖ , stᵍ ⟧ᶜ
-    dCert@(regpool c poolParams) ⟦ stᵈ , stᵖ' , stᵍ ⟧ᶜ (CERT-pool h)
+  Computational-CERT .completeness ⟦ _ , pp , _ , _ ⟧ᶜ ⟦ stᵈ , stᵖ , stᵍ ⟧ᶜˢ
+    dCert@(regpool c poolParams) ⟦ stᵈ , stᵖ' , stᵍ ⟧ᶜˢ (CERT-pool h)
     with computeProof pp stᵖ dCert | completeness _ _ _ _ h
   ... | just _ | refl = refl
-  Computational-CERT .completeness ⟦ _ , pp , _ ⟧ᶜ ⟦ stᵈ , stᵖ , stᵍ ⟧ᶜ
-    dCert@(retirepool c e) ⟦ stᵈ , stᵖ' , stᵍ ⟧ᶜ (CERT-pool h)
+  Computational-CERT .completeness ⟦ _ , pp , _ , _ ⟧ᶜ ⟦ stᵈ , stᵖ , stᵍ ⟧ᶜˢ
+    dCert@(retirepool c e) ⟦ stᵈ , stᵖ' , stᵍ ⟧ᶜˢ (CERT-pool h)
     with completeness _ _ _ _ h
   ... | refl = refl
-  Computational-CERT .completeness Γ ⟦ stᵈ , stᵖ , stᵍ ⟧ᶜ
+  Computational-CERT .completeness Γ ⟦ stᵈ , stᵖ , stᵍ ⟧ᶜˢ
     dCert@(regdrep c d an)
-    ⟦ stᵈ , stᵖ , stᵍ' ⟧ᶜ (CERT-vdel h)
+    ⟦ stᵈ , stᵖ , stᵍ' ⟧ᶜˢ (CERT-vdel h)
     with computeProof Γ stᵍ dCert | completeness _ _ _ _ h
   ... | just _ | refl = refl
-  Computational-CERT .completeness Γ ⟦ stᵈ , stᵖ , stᵍ ⟧ᶜ
-    dCert@(deregdrep c) ⟦ stᵈ , stᵖ , stᵍ' ⟧ᶜ (CERT-vdel h)
+  Computational-CERT .completeness Γ ⟦ stᵈ , stᵖ , stᵍ ⟧ᶜˢ
+    dCert@(deregdrep c) ⟦ stᵈ , stᵖ , stᵍ' ⟧ᶜˢ (CERT-vdel h)
     with computeProof Γ stᵍ dCert | completeness _ _ _ _ h
   ... | just _ | refl = refl
-  Computational-CERT .completeness Γ ⟦ stᵈ , stᵖ , stᵍ ⟧ᶜ
-    dCert@(ccreghot c mkh) ⟦ stᵈ , stᵖ , stᵍ' ⟧ᶜ (CERT-vdel h)
+  Computational-CERT .completeness Γ ⟦ stᵈ , stᵖ , stᵍ ⟧ᶜˢ
+    dCert@(ccreghot c mkh) ⟦ stᵈ , stᵖ , stᵍ' ⟧ᶜˢ (CERT-vdel h)
```

Comment:
I'm just learning how to do metaprogramming in Agda, so I'm not yet sure it's possible, but it would be nice to have type classes like `HasTernaryConstructor` (and `HasQuaternaryConstructor`).  Members of this type class would be any record type with three (resp., four) fields, and presumably we could then use instance resolution for the constructor, which we would call `⟦_,_,_⟧` (resp., `⟦_,_,_,_⟧`).  Then we could get rid of the superscripts and check off one more item in the PDF cleanup issue #146 .

If you have any ideas about whether this is possible and how we might accomplish this, please share.  Otherwise, I'll continue learning about Agda's metaprogramming capabilities (probably somewhat aimlessly, until someone points me in the right direction 😆).

Reply:
On the other hand, there's the default constructor `mk`.  Could we be using that in place of the bold brackets and superscripts?  Of course that would be less visually appealing than our brackets, especially for the pdf. Is it possible to rename `mk` generally, instead of on a case-by-case basis (when defining a particular record)?

Reply:
Yeah, that's the typeclass(es) that I've had in mind for #146. What you've described above should work fine. 

Metaprogramming isn't super helpful here, the only thing it could do would be to automatically generate the instances. But since those definitions should be simple one-liners anyway, there isn't much gained.

Reply:
Ah, but one problem with this is that it doesn't work for patterns, only for expressions. So this wouldn't actually help in this particular case, but since the STS constructors only pretend to pattern-match (but really are expressions) it works there. But Agda has some overloading with patterns, so maybe we could use overloaded patterns + some type classes to get the same effect.

Reply:
Indeed doing this generically for patterns probably is not possible with the current machinery, but for expressions meta-programming comes in handy for sure:
```agda
mkRecordⁿ : List Term → Hole → TC ⊤
mkRecordⁿ xs hole = do
  def n [] ← inferType hole
    where _ → error $ "[mkRecordⁿ] can only construct values of a defined name"
  record-type recCon _ ← getDefinition n
    where _ → error $ "[mkRecordⁿ] can only construct records types"
  unify hole $ con recCon (vArg <$> xs)
macro
  ⟦_⟧ = mkRecordⁿ

  _`∷_ : Term → List Term → Hole → TC ⊤
  (x `∷ y) hole = do
    `x ← quoteTC x
    `y ← quoteTC y
    unify hole $ con (quote _∷_) (vArg `x ∷ vArg `y ∷ [])
infixr 4 _`∷_

private
  record R : Set where
    field x : ℕ; s : String
  _ = R ∋ ⟦ 0 `∷ "zero" `∷ [] ⟧
  record Q : Set where
    field x : ℕ; s : String; y : ℕ; w : String
  _ = Q ∋ ⟦ 0 `∷ "zero" `∷ 1 `∷ "one" `∷ [] ⟧
```